### PR TITLE
[JIT] Add CUDNN Conv-Add-Relu fusion for Frozen Model Optimization

### PR DESCRIPTION
**Pull request by desertfire**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#52102 [JIT] Add CUDNN Conv-Add-Relu fusion for Frozen Model Optimization**

Summary: Fuse Conv-Relu based on cudnnConvolutionBiasActivationForward as a part of the Frozen Model Optimization. The optimization is not turned on as default yet.

Test Plan:
```
python test/test_jit.py -k test_freeze_conv_relu_fusion
```

Reviewers:

Subscribers:

Tasks: T78068152

Tags:

Differential Revision: [D26646100](https://our.internmc.facebook.com/intern/diff/D26646100)

### CODE REVIEW

File: test/jit/test_freezing.py
Comment (eellison) [inline test/jit/test_freezing.py:1566]: We're going to want to add more tests around bias being present/not present, and NCHW memory format, as well as Conv1d/Conv2d/Conv3d. This is a good example test: https://github.com/pytorch/pytorch/blob/master/test/jit/test_freezing.py#L1373.

**Review by eellison — COMMENTED** [review]
Great start 🚀 🚀 🚀 

I would say the big things we should figure out before landing is:
-  bias not present
- interaction with `torch.backends.cudnn.benchmark` - in the first version of this PR, just doing the default algorithm is fine, and we can integrate with the flag in a second PR
- We need to audit the scenarios under which pytorch eager [doesn't dispatch to cudnn](https://github.com/pytorch/pytorch/blob/0a8058469e1cd3642ab5a275512d8ed63cf163a1/aten/src/ATen/native/Convolution.cpp#L190). There may be cases where cudnn does not correctly compute things. Those cases that depend on the weight tensor we can avoid AOT, but we might need to do some runtime logic that depends on the input tensors.

File: aten/src/ATen/cudnn/Descriptors.h
Comment (eellison) [inline aten/src/ATen/cudnn/Descriptors.h:304]: Cool!! We should be able to use this in the future to support other activations as well https://docs.nvidia.com/deeplearning/cudnn/api/index.html#cudnnActivationMode_t.

File: aten/src/ATen/native/cudnn/Conv_v7.cpp
Comment (desertfire) [reply aten/src/ATen/native/cudnn/Conv_v7.cpp:871]: By calling "AlgoIterator<cudnnConvolutionFwdAlgoPerf_t>(args, benchmark)", we are leveraging https://github.com/pytorch/pytorch/blob/75cba9d0d18752bb800a1a7697976536590ccf5b/aten/src/ATen/native/cudnn/Conv_v7.cpp#L280 to fetch the best conv algorithm.

File: aten/src/ATen/native/cudnn/Conv_v7.cpp
Comment (desertfire) [reply aten/src/ATen/native/cudnn/Conv_v7.cpp:898]: CUDNN API doc says this is the full statement, y = act ( alpha1 * conv(x) + alpha2 * z + bias ); In our implementation, alpha1 is always 1, and alpha2 is the passed-in alpha parameter.

File: aten/src/ATen/native/cudnn/Conv_v7.cpp
Comment (eellison) [inline aten/src/ATen/native/cudnn/Conv_v7.cpp:885]: The equivalent pytorch eager code when we're not benchmarking is just:
```
    if (!benchmark) {
      AT_CUDNN_CHECK_WITH_SHAPES(cudnnGetConvolutionForwardAlgorithm_v7(
          args.handle,
          args.idesc.desc(),
          args.wdesc.desc(),
          args.cdesc.desc(),
          args.odesc.desc(),
          num_algos,
          &perf_count,
          perf_results.get()), args);
    } else { 
```

why are we doing the `try_all([&](const cudnnConvolutionFwdAlgoPerf_t` logic ?

File: test/jit/test_freezing.py
Comment (eellison) [inline test/jit/test_freezing.py:1750]: i would consider making `identity = torch.rand_like(x)` just to stress this a little more

File: torch/csrc/jit/passes/frozen_conv_relu_fusion.cpp
Comment (eellison) [inline torch/csrc/jit/passes/frozen_conv_relu_fusion.cpp:67]: we might consider making this so we only remove them if an input to add_ is Convolution and an input to relu_ is Convolution or an add_ which has a convolution input. (not necessary for the first PR)

File: torch/csrc/jit/passes/frozen_graph_optimizations.cpp
Comment (eellison) [inline torch/csrc/jit/passes/frozen_graph_optimizations.cpp:29]: I think we might want to integrate this into the main api only after a little more testing on benchmark suite and such.

File: torch/csrc/jit/passes/frozen_conv_relu_fusion.cpp
Comment (eellison) [inline torch/csrc/jit/passes/frozen_conv_relu_fusion.cpp:55]: leftover code

File: aten/src/ATen/native/cudnn/ConvShared.cpp
Comment (eellison) [inline aten/src/ATen/native/cudnn/ConvShared.cpp:505]: We should actually add this logic to the fusion pass bc it doesnt exist right now

File: torch/csrc/jit/passes/frozen_conv_relu_fusion.cpp
Comment (eellison) [inline torch/csrc/jit/passes/frozen_conv_relu_fusion.cpp:59]: if bias or `z` are present we might consider doing some error checking up front.

File: aten/src/ATen/native/cudnn/Conv_v7.cpp
Comment (eellison) [inline aten/src/ATen/native/cudnn/Conv_v7.cpp:837]: https://docs.nvidia.com/deeplearning/cudnn/api/index.html#cudnnActivationMode_t : ` cudnnConvolutionBiasActivationForward() function must use CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_​PRECOMP_GEMM` - i'm not sure we can implement benchmarking at all.

**Review by eellison — APPROVED** [review]
Looks great 😍 😍 😍 

I think we might consider turning it on in freezing as a separate PR just so this is less likely to break anything, and we can do a little more testing/benchmarking on a variety of models.

It would great if you would post a few benchmarks onto the pr description - i know you said up to around 3% when it was just Conv-Relu, not yet Conv-Add-Relu, i'd be curious about the latter case. Considering Conv-Relu and Conv-Add-Relu is like the entire compute pattern of many vision models after freezing, so we should definitely take whatever we can get.

It might be unlikely but we can keep our eyes out in the future for Convolution-Add - (Non pointwise op) and consider fusing with `CUDNN_ACTIVATION_IDENTITY`.

Tagging @ngimel for someone else familiar with Cudnn, would be great to get another set of eyes on it.

File: aten/src/ATen/native/cudnn/ConvShared.cpp
Comment (ngimel) [inline aten/src/ATen/native/cudnn/ConvShared.cpp:537]: you are setting bias to uninitialized output tensor here, how does this work?

File: aten/src/ATen/native/cudnn/Conv_v7.cpp
Comment (desertfire) [reply aten/src/ATen/native/cudnn/Conv_v7.cpp:837]: "Only the CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_​PRECOMP_GEMM algo is enabled with CUDNN_ACTIVATION_IDENTITY. In other words, in the cudnnActivationDescriptor_t structure of the input activationDesc, if the mode of the cudnnActivationMode_t field is set to the enum value CUDNN_ACTIVATION_IDENTITY, then the input cudnnConvolutionFwdAlgo_t of this function cudnnConvolutionBiasActivationForward() must be set to the enum value CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_​PRECOMP_GEMM. For more information, see cudnnSetActivationDescriptor()."  -- My interpretation is we can still do benchmarking if the activation function is RELU.

File: torch/csrc/jit/passes/frozen_conv_relu_fusion.cpp
Comment (eellison) [reply torch/csrc/jit/passes/frozen_conv_relu_fusion.cpp:67]: for (Value * v: node->inputs()) {
    if (v->node()->kind() == aten::conv2d) {
    
    }
}

File: test/jit/test_freezing.py
Comment (eellison) [inline test/jit/test_freezing.py:1775]: maybe add a test that checks if torch.jit.save will error (we dont want to have to maintain BC for this operator)

File: torch/csrc/jit/passes/frozen_graph_optimizations.cpp
Comment (desertfire) [reply torch/csrc/jit/passes/frozen_graph_optimizations.cpp:29]: Let me revert this one.

File: aten/src/ATen/native/cudnn/ConvShared.cpp
Comment (desertfire) [reply aten/src/ATen/native/cudnn/ConvShared.cpp:535]: I thought "auto output_t = at::native::empty_cuda(..." initializes output as zero. If it is not the right assumption, I will create zeros explicitly.

File: aten/src/ATen/native/cudnn/ConvShared.cpp
Comment (desertfire) [reply aten/src/ATen/native/cudnn/ConvShared.cpp:537]: I thought "auto output_t = at::native::empty_cuda(..." initializes output as zero. If it is not the right assumption, I will create zeros explicitly.

Comment (ngimel) [reply aten/src/ATen/native/cudnn/ConvShared.cpp:537]: It's not the right assumption, but ideally you should avoid explicitly creating tensor of zeros, and instead use corresponding arguments in cudnn call (like beta=0) to make sure those uninitialized values are never read and only written to.

**Pull request closed without merging**
